components/brave_news/browser/brave_news_controller.cc

```diff
@@ -108,20 +108,24 @@ BraveNewsController::BraveNewsController(
     favicon::FaviconService* favicon_service,
     brave_ads::AdsService* ads_service,
     history::HistoryService* history_service,
-    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    std::unique_ptr<DirectFeedFetcher::Delegate> direct_feed_fetcher_delegate)
     : favicon_service_(favicon_service),
       ads_service_(ads_service),
       api_request_helper_(GetNetworkTrafficAnnotationTag(), url_loader_factory),
       private_cdn_request_helper_(GetNetworkTrafficAnnotationTag(),
                                   url_loader_factory),
       history_service_(history_service),
       url_loader_factory_(url_loader_factory),
-      pref_manager_(*prefs),
-      news_metrics_(prefs, pref_manager_),
-      direct_feed_controller_(url_loader_factory),
       task_runner_(base::ThreadPool::CreateSingleThreadTaskRunner(
           {base::MayBlock(), base::TaskPriority::BEST_EFFORT,
            base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN})),
+      direct_feed_fetcher_delegate_(direct_feed_fetcher_delegate.release(),
+                                    base::OnTaskRunnerDeleter(task_runner_)),
```

Comment:
I don't understand why this should be destroyed on `task_runner_`. It's used only from `direct_feed_controller_` which doesn't depend on `task_runner_`.

Reply:
I thought the same at first but I think it is passed to |engine_| which does run on |task_runner|. Not sure of the implications - the stuff it uses seems to be threadsafe...

Reply:
See comment: https://github.com/brave/brave-core/pull/26849#discussion_r1870242452

Reply:
ah, I missed that the delegate is passed to the engine down below.

are you sure it's safe to call `ShouldUpgradeToHttps` on any thread?

Reply:
The methods it uses say they're Threadsafe in the comments but I was a little wary :D

Reply:
upon further inspection, it looks like the https exception service may not be thread safe. changed to call the method on the UI thread

Reply:
you're using it on UI thread now, but still destroying on a worker thread. 
Please remove this `OnTaskRunnerDeleter` and use `WeakPtr` instead of `Unretained` when you post a call on UI thread.

It doesn't matter at which thread you issue the weakptr, it matters where you dereference it (check the weak pointer and get the actual raw pointer from it) and destrory the object this weakptr operates on.

in your case the dereference happens on UI thread when the task you posted is about to be run.

Reply:
initially i tried using a weak ptr, but hit a wall when i discovered that weak ptr calls are not supported for non-void types. i used `PostTask` instead as a workaround, but the solution is less elegant. lmk what you think